File: leetCode/c/p36_ValidSudoku/lc36.c
/* Determine if a 9 x 9 Sudoku board is valid. 

Only the filled cells need to be validated according to the following rules:

    - Each row must contain the digits 1-9 without repetition.
    - Each column must contain the digits 1-9 without repetition.
    - Each of the nine 3 x 3 sub-boxes of the grid must contain the digits 1-9 without repetition.

Note:

A Sudoku board (partially filled) could be valid but is not necessarily solvable.
Only the filled cells need to be validated according to the mentioned rules. */

#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include "../../../mylib/mylib.h"

bool isNumberPossible(char** board, int y, int x) {

    int numOfInst = 0;
    int x0 = (x / 3) * 3;
    int y0 = (y / 3) * 3;

    for(int i=0; i<9; i++) {
        if(board[y][i] == board[y][x])
            numOfInst++; 
    }
    
    if(numOfInst > 1)
        return false;

    numOfInst = 0; 

    for(int i=0; i<9; i++) {
        if(board[i][x] == board[y][x])
            numOfInst++; 
    }
    
    if(numOfInst > 1)
        return false;
 
    numOfInst = 0;  

    for(int i=0; i<3; i++) {
        for(int j=0; j<3; j++) {
            if(board[y0+i][x0+j] == board[y][x])
                numOfInst++;
        }
    }

    if(numOfInst > 1)
        return false;

    return true;
}

bool isValidSudoku(char** board) {

    for(int i=0; i<9; i++) {
        for(int j=0; j<9; j++) {
            if(board[i][j] != '.' && isNumberPossible(board, i, j) == false) {
                return false;
            }
        }
    }

    return true;
}

int main(void) {

    // Test 1:
    printf("Test 1:\n\n");

    char* board1[9] = {"53..7....",
                       "6..195...",
                       ".98....6.",
                       "8...6...3",
                       "4..8.3..1",
                       "7...2...6",
                       ".6....28.",
                       "...419..5",
                       "....8..79"};

    printMatrixChar(board1, 9, 9);
    
    if(isValidSudoku(board1))
        printf("\nGiven board is valid!\n");
    else
        printf("\nGiven board is NOT valid!\n");

    // Test 2:
    printf("\nTest 2:\n\n");

    char* board2[9] = {"83..7....",
                       "6..195...",
                       ".98....6.",
                       "8...6...3",
                       "4..8.3..1",
                       "7...2...6",
                       ".6....28.",
                       "...419..5",
                       "....8..79"};

    printMatrixChar(board2, 9, 9);
    
    if(isValidSudoku(board2))
        printf("\nGiven board is valid!\n");
    else
        printf("\nGiven board is NOT valid!\n");

    return EXIT_SUCCESS; 
}

File: leetCode/python/p36_ValidSudoku/lc36.py
#!/usr/bin/python3

import numpy as np

class Solution:
    '''
        Determine if a 9 x 9 Sudoku board is valid. Only the filled cells need to be validated according to the following rules:

            - Each row must contain the digits 1-9 without repetition.
            - Each column must contain the digits 1-9 without repetition.
            - Each of the nine 3 x 3 sub-boxes of the grid must contain the digits 1-9 without repetition.
        Note:

            A Sudoku board (partially filled) could be valid but is not necessarily solvable.
            Only the filled cells need to be validated according to the mentioned rules.
    '''
    def print_board(self, board):
    
        print(np.matrix(board))

    def is_num_possible(self, y, x, board):

        num_of_inst = 0
        x0 = (x // 3) * 3
        y0 = (y // 3) * 3
        
        for i in range(9):
            if board[y][i] == board[y][x]:
                num_of_inst += 1

        if num_of_inst > 1:
            return False

        num_of_inst = 0

        for i in range(9):
            if board[i][x] == board[y][x]:
                num_of_inst += 1

        if num_of_inst > 1:
            return False

        num_of_inst = 0

        for i in range(3):
            for j in range(3):
                if board[y0+i][x0+j] == board[y][x]:
                    num_of_inst += 1

        if num_of_inst > 1:
            return False

        return True
    
    def isValidSudoku(self, board) -> bool:
        
        for i in range(9):
            for j in range(9):
                if board[i][j] != '.' and self.is_num_possible(i, j, board) == False:
                    return False

        return True

if __name__ == '__main__':

    s = Solution()
    
    # Test 1:
    
    board1 = [['5','3','.','.','7','.','.','.','.'],
              ['6','.','.','1','9','5','.','.','.'],
              ['.','9','8','.','.','.','.','6','.'],
              ['8','.','.','.','6','.','.','.','3'],
              ['4','.','.','8','.','3','.','.','1'],
              ['7','.','.','.','2','.','.','.','6'],
              ['.','6','.','.','.','.','2','8','.'],
              ['.','.','.','4','1','9','.','.','5'],
              ['.','.','.','.','8','.','.','7','9']]

    print('Test 1:\n')
    s.print_board(board1)

    if s.isValidSudoku(board1):
        print('\nGiven board is valid!')
    else:
        print('\nGiven board is NOT valid!')

    # Test 2:

    board2 = [['8','3','.','.','7','.','.','.','.'],
              ['6','.','.','1','9','5','.','.','.'],
              ['.','9','8','.','.','.','.','6','.'],
              ['8','.','.','.','6','.','.','.','3'],
              ['4','.','.','8','.','3','.','.','1'],
              ['7','.','.','.','2','.','.','.','6'],
              ['.','6','.','.','.','.','2','8','.'],
              ['.','.','.','4','1','9','.','.','5'],
              ['.','.','.','.','8','.','.','7','9']]

    print('\nTest 2:\n')
    s.print_board(board2)

    if s.isValidSudoku(board2):
        print('\nGiven board is valid!')
    else:
        print('\nGiven board is NOT valid!')
